src/main/kotlin/org/radarcns/gateway/kafka/AvroProcessor.kt

```diff
@@ -122,27 +121,24 @@ class AvroProcessor {
         }
     }
 
-    private class Auth(jwt: DecodedJWT) {
-        val projectIds: List<String>
+    private class Auth(jwt: RadarToken) {
+        val projectIds: Set<String>
         val userId: String = jwt.subject
         val sourceIds: Collection<String>
         init {
             // get roles as a list, empty if not set
-            val rolesClaim = jwt.getClaim(ROLES_CLAIM)?.asList(String::class.java)
-                    ?: Collections.emptyList()
+            val rolesClaim = jwt.roles ?: Collections.emptyMap()
 
-            projectIds = rolesClaim.stream()
-                    .filter { it.endsWith(PARTICIPANT_SUFFIX) }
-                    .map { it.substring(0, it.length - PARTICIPANT_SUFFIX.length) }
-                    .collect(Collectors.toList())
+            projectIds = rolesClaim.filterValues { it.contains(Util.PARTICIPANT_SUFFIX) }.toMap()
+                    .keys
+            jwt.hasPermissionOnProject(Permission.MEASUREMENT_CREATE , projectIds.first())
```

Comment:
perhaps it would be good to merge the above two statements:
```kotlin
rolesClaim.entries.filter {
    it.value.contains(Util.PARTICIPANT_SUFFIX)
            && jwt.hasPermissionOnProject(Permission.MEASUREMENT_CREATE, it.key)
    }.keys
```

Reply:
Or probably just leave out the `PARTICIPANT_SUFFIX` check, we just need to know that they are allowed to create measurements.